VACMS-9357 Shows facility hours in view display

omahane opened the pull request:
## Description

Relates to #9357
Only shows the hours on the view display.

## Testing done
Manually

## Screenshots


## QA steps

What needs to be checked to prove this works?  
What needs to be checked to prove it didn't break any related things?  
What variations of circumstances (users, actions, values) need to be checked?  

As user _uid_ with _user_role_
1. Do this
   - [ ] Validate that
2. Then
   - [ ] Validate that
3. Then validate Acceptance Criteria from issue
   - [ ] a
   - [ ] b
   - [ ] c

### Definition of Done

- [ ] Documentation has been updated, if applicable.
- [ ] Tests have been added if necessary.
- [ ] Automated tests have passed.
- [ ] Code Quality Tests have passed.
- [ ] Acceptance Criteria in related issue are met.
- [ ] Manual Code Review Approved.

### Select Team for PR review

- [ ] `Platform CMS Team`
- [ ] `Sitewide program`
- [ ] `⭐️ Sitewide CMS`
- [ ] `⭐️ Public websites`
- [ ] `⭐️ Facilities`
- [ ] `⭐️ User support`


### Is this PR blocked by another PR?

- [ ] `DO NOT MERGE`

### Does this PR need review from a Product Owner

- [ ] `Needs PO review`

### CMS user-facing annoucement

Is an announcement needed to let editors know of this change?
- [ ] Yes, and it's written in issue ____ and queued for publication.
  - [ ] Merge and ping the UX writer so they are ready to publish after deployment
- [ ] Yes, but it hasn't yet been written
  - [ ] Don't merge yet -- ping the UX writer to write and queue content
- [ ] No announcement is needed for this code change.
  - [ ] Merge & carry on unburdened by announcements

swirtSJW commented on docroot/modules/custom/va_gov_facilities/va_gov_facilities.module:
```diff
@@ -12,3 +12,34 @@ function va_gov_facilities_page_attachments(array &$attachments) {
+
+      // @todo will this always be a node?
+      $viewBuilder = \Drupal::entityTypeManager()->getViewBuilder('node');
+      $referenced_node = $node_parent->get('field_facility_location')->referencedEntities();
```
```suggestion
       //* @var /Drupal/kdkdks/field_list_items */
       $field_list_items= $node_parent->get('field_facility_location')
      $referenced_node =  $field_list_items->referencedEntities();
```

omahane replied:
I've refactored using the `@phpstan-ignore-next-line`

swirtSJW commented on docroot/modules/custom/va_gov_facilities/va_gov_facilities.module:
```diff
@@ -12,3 +12,34 @@ function va_gov_facilities_page_attachments(array &$attachments) {
+
+      // @todo will this always be a node?
+      $viewBuilder = \Drupal::entityTypeManager()->getViewBuilder('node');
+      $referenced_node = $node_parent->get('field_facility_location')->referencedEntities();
```
```suggestion
       // @phpstan-ignore-next-line
      $referenced_node = $node_parent->get('field_facility_location')->referencedEntities();
```

swirtSJW commented:
This is still in draft so this feedback may be premature.  However, I noticed what is being implemented does not match the Figma
![image](https://user-images.githubusercontent.com/5752113/186566670-3b8fc74c-b2bf-4fbd-ac58-314db0e4b3ee.png)

swirtSJW commented:
Also getting this error on trying to add a new facilty health service,
```
The website encountered an unexpected error. Please try again later.
Error: Call to a member function get() on null in _va_gov_facilities_create_custom_variable() (line 100 of modules/custom/va_gov_facilities/va_gov_facilities.module).
```
https://pr10361-bpic1dvo2pyvombh2x6xta0dvdgxqhnk.ci.cms.va.gov/node/add/health_care_local_health_service
![image](https://user-images.githubusercontent.com/5752113/186567170-39e2ae14-c5e8-4629-a44d-bed29f65ab36.png)

omahane commented:
@swirtSJW Also, I think the design shows edit display, rather than view display. In fact, I don't think there is a design for the view display.

A couple questions, the design shows radio buttons, but the form display is currently a dropdown. Should this change, too, then? I did try to change it and it broke the current current behavior of showing the hours form when "Provide specific hours for this service" was chosen.

There's an option to add another service location but I can't find any examples of multiple service locations. I expect that we anticipate having multiple locations someday, but do we have them now?

tjheffner commented on docroot/modules/custom/va_gov_facilities/va_gov_facilities.libraries.yml:
```diff
@@ -2,3 +2,8 @@ facility_status_details:
   version: 1.x
   js:
     js/facility_status_details.js: {}
+# use_facility_hours:
```
commented out code here

tjheffner commented on docroot/modules/custom/va_gov_backend/js/service_location_hours.es6.js:
```diff
@@ -2,35 +2,82 @@
+      if (toggle.value === value) {
+        table.style.display = "block";
+        if (value === "0") {
+          $(table)
```
probably don't/shouldn't need to introduce jquery as a dependency here

tjheffner reviewed (commented):
overall this is looking/feeling much better. a few comments on things i noticed, feel free to disregard if i'm totally off-base :)

when toggling the between the hours radio buttons on `/node/add/health_care_local_health_service` i see some errors pop up in the console
![Screen Shot 2022-09-01 at 9 02 03 AM](https://user-images.githubusercontent.com/11279744/187960518-1653b17f-a9d0-4b99-8acc-2f1845e96e7a.png)

not sure how critical that is, i do not see those console errors on an existing VAMC Facility Health Service, e.g. `/node/11981/edit`

omahane commented:
@swirtSJW Assuming the tests pass, this is ready to review.

swirtSJW commented on docroot/modules/custom/va_gov_facilities/va_gov_facilities.module:
```diff
@@ -12,3 +19,96 @@ function va_gov_facilities_page_attachments(array &$attachments) {
+ *   Name of the field to.
+ * @param array $haystack_bundles
+ *   Names of bundles to compare against the type of node we have.
+ */
```
Missing @return being defined.

swirtSJW commented on docroot/modules/custom/va_gov_facilities/va_gov_facilities.module:
```diff
@@ -12,3 +19,96 @@ function va_gov_facilities_page_attachments(array &$attachments) {
+}
+
+/**
+ * Tests to see if we should preprocess.
```
This says it is doing a test, but it is really getting a node.  And the function itself is named in a way that I can't tell what it does
"_va_gov_facilities_test_preprocess_show"  is test the verb or is show?

swirtSJW commented on docroot/modules/custom/va_gov_facilities/va_gov_facilities.module:
```diff
@@ -12,3 +19,96 @@ function va_gov_facilities_page_attachments(array &$attachments) {
+ */
+function _va_gov_facilities_test_preprocess_show($test_object, string $field_name, array $haystack_bundles) {
+  $field_value = $test_object->get($field_name)->value;
+  $node_parent = _vagovclaro_get_node_parent($test_object);
```
This is a function that is intentionally classified by its _ prefix as an internal function, meaning it is only supposed to be called by its own module (vagovclaro), but here we are violating that by calling it from another module.  Granted the underscore is only adopted convention, but it is still kind of a risky practice, especially because it is creating an undeclared dependency on the another module/theme.

swirtSJW commented on docroot/modules/custom/va_gov_facilities/va_gov_facilities.module:
```diff
@@ -12,3 +19,96 @@ function va_gov_facilities_page_attachments(array &$attachments) {
+ * @param string $field_to_render
+ *   Field we want to render.
+ */
+function _va_gov_facilities_create_custom_variable($node, $related_field, $field_to_render) {
```
Function  name and descripton are misleading 
_va_gov_facilities_create_custom_variable VS Creates custom array for render array

swirtSJW commented on docroot/modules/custom/va_gov_facilities/va_gov_facilities.module:
```diff
@@ -12,3 +19,96 @@ function va_gov_facilities_page_attachments(array &$attachments) {
+/**
+ * Creates custom array for render array.
+ *
+ * @param object $node
```
If something is a node, it should be hinted as a nodeinterface.

swirtSJW commented on docroot/modules/custom/va_gov_facilities/va_gov_facilities.module:
```diff
@@ -12,3 +19,96 @@ function va_gov_facilities_page_attachments(array &$attachments) {
+ * @param string $related_field
+ *   Field that relates our node to another.
+ * @param string $field_to_render
+ *   Field we want to render.
```
missing @return

swirtSJW commented on docroot/modules/custom/va_gov_facilities/va_gov_facilities.module:
```diff
@@ -12,3 +19,96 @@ function va_gov_facilities_page_attachments(array &$attachments) {
+    $viewBuilder = \Drupal::entityTypeManager()->getViewBuilder('node');
+    $output = $viewBuilder->viewField($value, 'full');
+    return $output;
+  }
```
This function only sometimes has a return?  THat is a risky pattern.

omahane commented:
@swirtSJW I removed the function I didn't really need (getting the node's parent). I moved the theme function to the backend and changed all the references. The theme already had a dependency on va_gov_backend. I didn't change the JS, as that might be a bigger undertaking.

swirtSJW commented on docroot/modules/custom/va_gov_backend/va_gov_backend.module:
```diff
@@ -2012,3 +2012,28 @@ function va_gov_backend_theme($existing, $type, $theme, $path) {
+ * @return \Drupal\node\Entity\Node|null
+ *   If the paragraph is attached to a node, return the node, otherwise NULL.
+ */
+function va_gov_backend_get_node_parent(Paragraph $paragraph) {
```
```suggestion
function va_gov_backend_get_parent_node(Paragraph $paragraph) {
```

swirtSJW commented on docroot/modules/custom/va_gov_backend/va_gov_backend.module:
```diff
@@ -2012,3 +2012,28 @@ function va_gov_backend_theme($existing, $type, $theme, $path) {
+  /** @var Drupal\paragraphs\Entity\Paragraph $parent */
+  $parent = $paragraph->getParentEntity();
+  if ($parent !== NULL && $parent->getEntityType()->id() === 'paragraph') {
+    $parent_node = va_gov_backend_get_node_parent($parent);
```
```suggestion
    $parent_node = va_gov_backend_get_parent_node($parent);
```

swirtSJW commented on docroot/modules/custom/va_gov_facilities/va_gov_facilities.module:
```diff
@@ -12,3 +19,22 @@ function va_gov_facilities_page_attachments(array &$attachments) {
+  $paragraph = $item_list->getEntity();
+  $field_name = "field_hours";
+  if ($paragraph->$field_name->value === $use_facility_hours) {
+    $node_parent = va_gov_backend_get_node_parent($paragraph);
```
```suggestion
    $node_parent = va_gov_backend_get_parent_node($paragraph);
```

swirtSJW commented on docroot/themes/custom/vagovclaro/vagovclaro.theme:
```diff
@@ -108,7 +107,7 @@ function vagovclaro_theme_suggestions_paragraph_alter(&$suggestions, $variables)
   /** @var \Drupal\paragraphs\Entity\Paragraph $paragraph */
   $paragraph = $variables['paragraph'];
-  $node_parent = _vagovclaro_get_node_parent($paragraph);
+  $node_parent = va_gov_backend_get_node_parent($paragraph);
```
```suggestion
  $node_parent = va_gov_backend_get_parent_node($paragraph);
```

swirtSJW reviewed (commented):
Set of changes to improve readability and accuracy.

swirtSJW commented on docroot/modules/custom/va_gov_facilities/src/EventSubscriber/FacilitiesSubscriber.php:
```diff
@@ -69,9 +72,38 @@ public static function getSubscribedEvents(): array {
   }
 
+  /**
+   * Creates custom array for render array.
```
```suggestion
   * Creates custom render array from field on referenced entity.
```

swirtSJW commented on docroot/modules/custom/va_gov_facilities/src/EventSubscriber/FacilitiesSubscriber.php:
```diff
@@ -69,9 +72,38 @@ public static function getSubscribedEvents(): array {
+   * @return array
+   *   Render array
+   */
+  public static function createCustomArray(NodeInterface $node, $related_field, $field_to_render) {
```
```suggestion
  public static function createRenderArrayFromFieldOnRefdEntity(NodeInterface $node, $related_field, $field_to_render) {
```
Specificity improves maintainability for our future selves.

swirtSJW commented on docroot/modules/custom/va_gov_facilities/src/EventSubscriber/FacilitiesSubscriber.php:
```diff
@@ -94,6 +126,46 @@ public function entityUpdate(EntityUpdateEvent $event): void {
+      $field_to_render = "field_office_hours";
+      $widget_items = $widget_complete_form['widget'];
+      for ($i = 0; isset($widget_items[$i]); $i++) {
+        $widget_complete_form['widget'][$i]['subform']['field_hours']['facility_hours'] = $this->createCustomArray($node, $related_field, $field_to_render);
```
```suggestion
        $widget_complete_form['widget'][$i]['subform']['field_hours']['facility_hours'] = $this->createRenderArrayFromFieldOnRefdEntity($node, $related_field, $field_to_render);
```

swirtSJW commented on docroot/modules/custom/va_gov_facilities/src/EventSubscriber/FacilitiesSubscriber.php:
```diff
@@ -69,9 +72,38 @@ public static function getSubscribedEvents(): array {
+   */
+  public static function createCustomArray(NodeInterface $node, $related_field, $field_to_render) {
+    // @phpstan-ignore-next-line Test is unclear what get() we are using
+    $referenced_node = $node->get($related_field)->referencedEntities();
```
```suggestion
    $referenced_entities = $node->get($related_field)->referencedEntities();
```
Misleading.  No reason to believe it is  a node. And inadvertently switched from plural to singular.

swirtSJW commented on docroot/modules/custom/va_gov_facilities/src/EventSubscriber/FacilitiesSubscriber.php:
```diff
@@ -69,9 +72,38 @@ public static function getSubscribedEvents(): array {
+    // @phpstan-ignore-next-line Test is unclear what get() we are using
+    $referenced_node = $node->get($related_field)->referencedEntities();
+    $output = [];
+    if (isset($referenced_node[0])) {
+      $node_entity = $referenced_node[0];
```
```suggestion
    if (isset($referenced_entities[0])) {
      $referenced_entity = $referenced_entities[0];
```

swirtSJW commented on docroot/modules/custom/va_gov_facilities/src/EventSubscriber/FacilitiesSubscriber.php:
```diff
@@ -69,9 +72,38 @@ public static function getSubscribedEvents(): array {
+    $output = [];
+    if (isset($referenced_node[0])) {
+      $node_entity = $referenced_node[0];
+      $value = $node_entity->get($field_to_render);
```
```suggestion
      $value = $referenced_entity->get($field_to_render);
```

swirtSJW commented on docroot/modules/custom/va_gov_facilities/src/EventSubscriber/FacilitiesSubscriber.php:
```diff
@@ -69,9 +72,38 @@ public static function getSubscribedEvents(): array {
+      $value = $node_entity->get($field_to_render);
+      $output['content']['weight'] = 10;
+      $output['#cache']['tags'] = $node_entity->getCacheTags();
+      $viewBuilder = \Drupal::entityTypeManager()->getViewBuilder('node');
```
Should be Dependency Injection, but because it is a static function, $this is not available.   So this is OK

swirtSJW replied:
This assumes the type is node, when there is nothing higher up that tests that assumption.  Would be safer to get the type from the entity.
```suggestion
      $viewBuilder = \Drupal::entityTypeManager()->getViewBuilder($referenced_entity->getEntityType());
```

swirtSJW commented on docroot/modules/custom/va_gov_facilities/src/EventSubscriber/FacilitiesSubscriber.php:
```diff
@@ -69,9 +72,38 @@ public static function getSubscribedEvents(): array {
+      $node_entity = $referenced_node[0];
+      $value = $node_entity->get($field_to_render);
+      $output['content']['weight'] = 10;
+      $output['#cache']['tags'] = $node_entity->getCacheTags();
```
```suggestion
      $output['#cache']['tags'] = $referenced_entity->getCacheTags();
```

swirtSJW reviewed (commented):
Renaming some misleading bits.  I'll take care of committing these.

swirtSJW commented on docroot/modules/custom/va_gov_facilities/src/EventSubscriber/FacilitiesSubscriber.php:
```diff
@@ -94,6 +126,46 @@ public function entityUpdate(EntityUpdateEvent $event): void {
+    $paragraph_entity_reference_field_name = $field_definition->getName();
+    if ($paragraph_entity_reference_field_name === "field_service_location") {
+      $item_list = $context['items'];
+      $node = $item_list->getEntity();
```
There is an assumption here that this is a node, and then passes the assumption to the static method that requires a nodeinterface.

swirtSJW commented on docroot/modules/custom/va_gov_facilities/src/EventSubscriber/FacilitiesSubscriber.php:
```diff
@@ -94,6 +126,46 @@ public function entityUpdate(EntityUpdateEvent $event): void {
+    if ($paragraph_entity_reference_field_name === "field_service_location") {
+      $item_list = $context['items'];
+      $node = $item_list->getEntity();
+      $related_field = "field_facility_location";
+      $field_to_render = "field_office_hours";
+      $widget_items = $widget_complete_form['widget'];
+      for ($i = 0; isset($widget_items[$i]); $i++) {
+        $widget_complete_form['widget'][$i]['subform']['field_hours']['facility_hours'] = $this->createRenderArrayFromFieldOnRefdEntity($node, $related_field, $field_to_render);
+      }
```
```suggestion
      if ($node instanceof NodeInterface) {
        $related_field = "field_facility_location";
        $field_to_render = "field_office_hours";
        $widget_items = $widget_complete_form['widget'];
        for ($i = 0; isset($widget_items[$i]); $i++) {
          $widget_complete_form['widget'][$i]['subform']['field_hours']['facility_hours'] = $this->createRenderArrayFromFieldOnRefdEntity($node, $related_field, $field_to_render);
        }
      }
```

swirtSJW reviewed (commented):
Some additional changes I will commit.

swirtSJW commented on docroot/modules/custom/va_gov_facilities/src/EventSubscriber/FacilitiesSubscriber.php:
```diff
@@ -94,6 +126,46 @@ public function entityUpdate(EntityUpdateEvent $event): void {
+  /**
+   * Widget complete form Event call.
+   *
+   * @param \Drupal\core_event_dispatcher\Event\Field\WidgetCompleteFormAlterEvent $event
+   *   The event.
+   */
+  public function widgetCompleteFormAlter(WidgetCompleteFormAlterEvent $event): void {
```
```suggestion
   * @param \Drupal\field_event_dispatcher\Event\Field\WidgetCompleteFormAlterEvent $event
   *   The event.
   */
  public function widgetCompleteFormAlter(WidgetCompleteFormAlterEvent $event): void {
```

swirtSJW reviewed (commented):
What should be the final round of changes.  I will commit.

swirtSJW commented:
Error: Call to undefined function va_gov_backend_get_node_parent() in va_gov_facilities_preprocess_field__paragraph__field_hours() (line 33 of modules/custom/va_gov_facilities/va_gov_facilities.module).

swirtSJW commented:
Found it  ^^ just missed committing one of my suggestions.

swirtSJW commented on docroot/modules/custom/va_gov_facilities/va_gov_facilities.module:
```diff
@@ -12,3 +19,22 @@ function va_gov_facilities_page_attachments(array &$attachments) {
+    if ($node_parent) {
+      $related_field = "field_facility_location";
+      $field_to_render = "field_office_hours";
+      $variables['actual_hours'] = FacilitiesSubscriber::createCustomArray($node_parent, $related_field, $field_to_render);
```
```suggestion
      $variables['actual_hours'] = FacilitiesSubscriber::createRenderArrayFromFieldOnRefdEntity($node_parent, $related_field, $field_to_render);
```
Bah.  I missed one.

swirtSJW commented on docroot/modules/custom/va_gov_facilities/src/EventSubscriber/FacilitiesSubscriber.php:
```diff
@@ -69,9 +72,38 @@ public static function getSubscribedEvents(): array {
+    if (isset($referenced_entities[0])) {
+      $referenced_entity = $referenced_entities[0];
+      $value = $referenced_entity->get($field_to_render);
+      $viewBuilder = \Drupal::entityTypeManager()->getViewBuilder($referenced_entity->getEntityType());
```
```suggestion
      $viewBuilder = \Drupal::entityTypeManager()->getViewBuilder($referenced_entity->getEntityTypeId());
```

swirtSJW commented:
Blarg!!  reminder to self: I will not refactor from the github UI, I will not I will not.

swirtSJW commented on docroot/modules/custom/va_gov_facilities/src/EventSubscriber/FacilitiesSubscriber.php:
```diff
@@ -69,9 +72,38 @@ public static function getSubscribedEvents(): array {
+      $value = $referenced_entity->get($field_to_render);
+      $viewBuilder = \Drupal::entityTypeManager()->getViewBuilder($referenced_entity->getEntityTypeId());
+      $output = $viewBuilder->viewField($value, 'full');
+      $output['content']['weight'] = 10;
```
I am seeing this warning when testing editing
![image](https://user-images.githubusercontent.com/5752113/191947214-96f58271-f416-488c-be12-d5251bd8bd7f.png)

Is the weight needed?

swirtSJW replied:
```suggestion
      $output['content']['#weight'] = 10;
```

swirtSJW reviewed (changes requested):
This all tests out good with the exception of the weight warning (see below) We need to address that.

swirtSJW reviewed (commented):
Approved.  Tests out great.
![image](https://user-images.githubusercontent.com/5752113/191949124-7e1edf02-b7da-483b-8516-e7876d69e488.png)


![image](https://user-images.githubusercontent.com/5752113/191949157-bfbce9de-1bb7-4052-9f10-e7831e126317.png)

swirtSJW commented:
Amazing job on the QA steps.  Thank you.   And great job sticking with this slog of a ticket.  We now have a great set of tools to do similar things (grabbing content from a referenced entity.)

swirtSJW commented:
![image](https://user-images.githubusercontent.com/5752113/192008201-d81015f7-5ed4-440b-95b1-f690e211873d.png)
